substrate/frame/staking/src/pallet/mod.rs

```diff
@@ -735,6 +743,41 @@ pub mod pallet {
 	#[pallet::storage]
 	pub(crate) type ChillThreshold<T: Config> = StorageValue<_, Percent, OptionQuery>;
 
+	/// Voter snapshot progress status.
+	///
+	/// If the status is `Ongoing`, it keeps a cursor of the last voter retrieved to proceed when
+	/// creating the next snapshot page.
+	#[pallet::storage]
+	pub(crate) type VoterSnapshotStatus<T: Config> =
+		StorageValue<_, SnapshotStatus<T::AccountId>, ValueQuery>;
+
+	/// Target snapshot progress status.
+	///
+	/// If the status is `Ongoing`, it keeps track of the last target account returned in the
+	/// snapshot.
+	#[pallet::storage]
+	pub(crate) type TargetSnapshotStatus<T: Config> =
+		StorageValue<_, SnapshotStatus<T::AccountId>, ValueQuery>;
+
+	/// Keeps track of an ongoing multi-page election solution request.
+	///
+	/// Stores the block number of when the first election page was requested. `None` indicates
+	/// that the election results haven't started to be fetched.
+	#[pallet::storage]
+	pub(crate) type ElectingStartedAt<T: Config> = StorageValue<_, BlockNumberFor<T>, OptionQuery>;
+
+	/// A bounded list of the "electable" stashes that resulted from a successful election.
+	#[pallet::storage]
+	pub(crate) type ElectableStashes<T: Config> =
+		StorageValue<_, BoundedBTreeSet<T::AccountId, T::MaxValidatorSet>, ValueQuery>;
+
+	/// Lock state for election data mutations.
+	///
+	/// While the lock is set, there should be no mutations on the ledgers/staking data, ensuring
+	/// that the data provided to [`Config::ElectionDataProvider`] is stable during all pages.
+	#[pallet::storage]
+	pub(crate) type ElectionDataLock<T: Config> = StorageValue<_, (), OptionQuery>;
```

Comment:
Do we need this? if someone changes their stake mid snapshot, what could go wrong?

Reply:
Okay the only thing is that we should be worried about is removing nodes form teh  `VoteList` during this, so that the iteration can resume at a reasonable spot. 

To be specific, we can remove anyone, other than the single poor voter/target that is the pointer to the next page. 

I suggest we add this logic directly to `VoterList` via some hook, and any operation like `chill` that attempts to remove a node might fail on `T::VoterList::remove()?`, so no extra code is needed here to handle the error. 

We certainly won't need a storage item here. We just need to communicate to the `VoterList` like:

```
/// Voter list pallet needs a single storage item, that is one `node` that cannot be removed. 
#[pallet::storage]
pub type IterationLock = StoarageValue<_>


/// set/uptess `who` as the lock prevents them from being removed from the list
/// This should be called per non-terminal page of snapshot creation
T::VoterList::set_lock(who)
/// Remove any existing locks
/// This is called on the terminal (last) page of the snapshot creation. 
T::VoterList::remove_lock()

Reply:
This code is from a previous version, thanks for spotting.

I agree, I added a new issue for this. Related to https://github.com/paritytech/polkadot-sdk/pull/6034#discussion_r1838262185.